src/mcp/client/auth.py

```diff
@@ -0,0 +1,346 @@
+"""
+OAuth 2.0 Client Implementation
+
+This module provides a complete OAuth 2.0 client implementation supporting:
+- Authorization Code Flow with PKCE
+- Dynamic Client Registration
+- Token Refresh
+- OAuth Server Metadata Discovery
+"""
+
+from typing import Protocol, cast
+from urllib.parse import urlencode, urljoin
+
+import httpx
+from pkce import generate_pkce_pair  # type: ignore
+
+from mcp.shared.auth import (
+    OAuthClientInformation,
+    OAuthClientInformationFull,
+    OAuthClientMetadata,
+    OAuthMetadata,
+    OAuthToken,
+)
+from mcp.types import LATEST_PROTOCOL_VERSION
+
+
+class OAuthClientProvider(Protocol):
+    """Protocol defining the interface for OAuth client providers."""
+
+    def get_redirect_url(self) -> str:
+        """Get the URL where the user agent will be redirected after authorization."""
+        ...
+
+    def get_client_metadata(self) -> OAuthClientMetadata:
+        """Get the metadata for this OAuth client."""
+        ...
+
+    def get_client_information(self) -> OAuthClientInformation | None:
+        """Get the client information as registered with the server.
+
+        Returns None if the client is not registered.
+        """
+        ...
+
+    def save_client_information(self, client_information: OAuthClientInformationFull):
+        """Optional Function to save the client information received from the server.
+
+        If implemented, this provider will support dynamic client registration.
+        """
+        ...
+
+    def get_token(self) -> OAuthToken | None:
+        """Get any existing OAuth tokens for the current session."""
+        ...
+
+    def save_token(self, token: OAuthToken):
+        """Save the new OAuth token after successful authorization."""
+        ...
+
+    def redirect_to_authorization(self, authorization_url: str):
+        """Redirect the user agent to begin the authorization flow."""
+        ...
+
+    def get_code_verifier(self) -> str:
+        """Get the PKCE code verifier for the current session."""
+        ...
+
+    def save_code_verifier(self, pkce_code_verifier: str):
+        """Save the PKCE code verifier before redirecting to authorization."""
+        ...
+
+
+class OAuthAuthorization:
+    """Main class for handling OAuth 2.0 authorization flows.
+
+    This class implements the OAuth 2.0 Authorization Code Flow with PKCE,
```

Comment:
Might be good to clarify this comment - this class actually uses the `provider` to try to authenticate and falls back to Oauth 2.0 w/ PKCE right?